include/deal.II/lac/trilinos_tpetra_sparse_matrix.templates.h

```diff
@@ -0,0 +1,661 @@
+        {
+          matrix->resumeFill();
+          compressed = false;
+        }
+
+      matrix->setAllToScalar(d);
+
+      return *this;
+    }
+
+
+
+    template <typename Number, typename Node>
+    SparseMatrix<Number, Node> &
+    SparseMatrix<Number, Node>::operator*=(const Number a)
+    {
+      matrix->scale(a);
+      return *this;
+    }
+
+
+
+    template <typename Number, typename Node>
+    SparseMatrix<Number, Node> &
+    SparseMatrix<Number, Node>::operator/=(const Number a)
+    {
+      Assert(a != 0, ExcDivideByZero());
+
+      const Number factor = 1.0 / a;
+      matrix->scale(factor);
+      return *this;
+    }
+
+
+
+    template <typename Number, typename Node>
+    void
+    SparseMatrix<Number, Node>::add(const size_type       row,
+                                    const size_type       n_cols,
+                                    const size_type      *col_indices,
+                                    const TrilinosScalar *values,
+                                    const bool            elide_zero_values,
+                                    const bool /*col_indices_are_sorted*/)
+    {
+      AssertIndexRange(row, this->m());
+
+      // If the matrix is marked as compressed, we need to
+      // call resumeFill() first.
+      if (compressed || matrix->isFillComplete())
+        {
+          matrix->resumeFill();
+          compressed = false;
+        }
+
+      // count zero entries;
+      size_t n_zero_entries = 0;
+      if (elide_zero_values)
+        for (size_t i = 0; i < n_cols; ++i)
+          if (values[i] == 0)
+            ++n_zero_entries;
+
+      // Convert the input into Teuchos::Array
+      Teuchos::Array<types::signed_global_dof_index> col_indices_array(
+        n_cols - n_zero_entries);
+      Teuchos::Array<Number> values_array(n_cols - n_zero_entries);
+      if (elide_zero_values)
+        {
+          size_t n_columns = 0;
+          for (size_t i = 0; i < n_cols; ++i)
+            {
+              // skip all zero entries, while filling the
+              if (values[i] != 0)
+                {
+                  AssertIsFinite(values[i]);
+                  AssertIndexRange(col_indices[i], n());
+                  AssertIndexRange(n_columns, n_zero_entries);
+                  col_indices_array[n_columns] = col_indices[i];
+                  values_array[n_columns]      = values[i];
+                  ++n_columns;
+                }
+            }
+          // Exit early if there is nothing to do
+          if (n_columns == 0)
+            return;
+        }
+      else
+        for (size_t i = 0; i < n_cols; ++i)
+          {
+            AssertIsFinite(values[i]);
+            AssertIndexRange(col_indices[i], n());
+            col_indices_array[i] = col_indices[i];
+            values_array[i]      = values[i];
+          }
+
+      // Sum the values into the global matrix.
+      matrix->sumIntoGlobalValues(row, col_indices_array(), values_array());
```

Comment:
The syntax `col_indices_array()` has been bugging me for the past several minutes, and I felt the need to look up what it does. Here, `operator()` extracts a view, which I find a rather incomprehensible way to do that. Would you mind changing this to the following, to make this clearer? Either the implicit conversion
```suggestion
      matrix->sumIntoGlobalValues(row, col_indices_array, values_array);
```
should work, or use
```suggestion
      matrix->sumIntoGlobalValues(row, col_indices_array.view(0,col_indices_array.size()), 
                         values_array.view(0, values_array.size()));
```
Presumably, the first is more easily readable.

Reply:
The first one should work.